lib/ui/painting/image_decoder.cc

```diff
@@ -57,30 +62,91 @@ static SkISize GetResizedDimensions(SkISize current_size,
 }
 
 static sk_sp<SkImage> ResizeRasterImage(sk_sp<SkImage> image,
+                                        std::unique_ptr<SkCodec> codec,
                                         std::optional<uint32_t> target_width,
                                         std::optional<uint32_t> target_height,
                                         const fml::tracing::TraceFlow& flow) {
   FML_DCHECK(!image->isTextureBacked());
 
-  const auto resized_dimensions =
+  TRACE_EVENT0("flutter", __FUNCTION__);
+  flow.Step(__FUNCTION__);
+
+  if (!target_width && !target_height) {
+    // The resized dimesions are the same as the intrinsic dimensions of the
+    // image. There is nothing to do.
+    return image->makeRasterImage();
+  }
+
+  auto resized_dimensions =
       GetResizedDimensions(image->dimensions(), target_width, target_height);
 
   if (resized_dimensions.isEmpty()) {
     FML_LOG(ERROR) << "Could not resize to empty dimensions.";
     return nullptr;
   }
 
-  if (resized_dimensions == image->dimensions()) {
-    // The resized dimesions are the same as the intrinsic dimensions of the
-    // image. There is nothing to do.
-    return image;
+  // This is where the scaling happens.
+  if (resized_dimensions.width() > image->dimensions().width() ||
+      resized_dimensions.height() > image->dimensions().height()) {
+    FML_LOG(WARNING)
+        << "Image is being upsized. Using cache(Height|Width) correctly?";
+    // TOOD(jeltsch): consider exiting here, there's no good reason to support
+    // upsampling in a "caching"-optimiztion context..
   }
 
-  TRACE_EVENT0("flutter", __FUNCTION__);
-  flow.Step(__FUNCTION__);
+  const bool aspect_ratio_changed =
+      std::abs(AspectRatio(resized_dimensions) -
+               AspectRatio(image->dimensions())) > 0.01;
+  if (aspect_ratio_changed) {
+    // This is probably a bug. If a user passes dimensions that change the
+    // aspect ratio in a "caching" context that's probably not working as
+    // intended and rather a signal that the API is hard to use.
+    FML_LOG(WARNING) << "Aspect ratio is being changed.";
+  }
```

Comment:
I don't think this is an actionable log for a user.

I think we should just preserve the aspect ratio (preserving the smaller side and shrinking the larger side) here and drop the warning. If someone actually wants a resizing API that does not preserve aspect ratio, we should add that later.

Reply:
Thanks for the review.

Just to avoid us bouncing back and forth, I like your suggestion. I've been going back and forth on this myself, i.e. but ended up preserving the current semantics to make the adoption simpler.

I'm all for changing the cache(Height|Width) API on the network image. It's really hard to use, i.e. accidentally changing aspect ratio or upsampling the image. I just didn't feel like folding this into my change, i.e. I wanted to keep it a strict improvement.

That said, if we're departing from this and feel like we should change the semantics, let's make sure we're all on the same page about this (@mehmetf).

Possible options are:
 1. as you suggested here, drop aspect ratio changes
 2. drop resizes that would increase the image size
 3. don't strictly respect cache(Width|Height) and don't resize after decoding at a smaller size, e.g. if target decoding size is within X% of target size, just leave it at decode size.

Let me know how far you think we should depart from the current semantics as part of the immediate change.
Keep in mind, that we can always change the semantics and/or even change the cache(Width|Height) APIs in follow-ups.
I trust your judgement, just let me know and I'll change the PR asap.

Reply:
I think all three of your options are worth exploring, and they don't necessarily have to be in this patch.

For aspect ratio _specifically_, we should either fix it in this patch or drop the log warning. Logging that a user is doing something broken with no way to fix it goes against our style guide.  

Perhaps we could add a parameter like `preserve_aspect_ratio` and pipe that up through the Dart interfaces as well.  Again, that's fine if that's in a separate patch, but I don't think we should introduce the new log here - it's more likely to cause confusion that the user is doing something wrong with no clear way to fix it.

Reply:
Ok, I suggest we don't fold semantic/behavioral changes into this change just to keep it simple.  I filed https://github.com/flutter/flutter/issues/48885